epochStart/metachain/rewardsStakingProvider.go

```diff
@@ -0,0 +1,165 @@
+package metachain
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math"
+	"math/big"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/epochStart"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const conversionBase = 10
+
+type ownerStats struct {
+	numEligible int
+	topUpValue  *big.Int
+}
+
+type rewardsStakingProvider struct {
+	mutCache sync.Mutex
+	cache    map[string]*ownerStats
+	systemVM vmcommon.VMExecutionHandler
+}
+
+// NewRewardsStakingProvider will create a new instance of a staking data provider able to aid in the final rewards
+// computation as this will retrieve the staking data from the system VM
+func NewRewardsStakingProvider(
+	systemVM vmcommon.VMExecutionHandler,
+) (*rewardsStakingProvider, error) {
+	//TODO make vmcommon.VMExecutionHandler implement NilInterfaceChecker
+	if check.IfNilReflect(systemVM) {
+		return nil, epochStart.ErrNilSystemVmInstance
+	}
+
+	rsp := &rewardsStakingProvider{
+		systemVM: systemVM,
+	}
+	rsp.Clean()
+
+	return rsp, nil
+}
+
+// Clean will reset the inner state of the called instance
+func (rsp *rewardsStakingProvider) Clean() {
+	rsp.mutCache.Lock()
+	rsp.cache = make(map[string]*ownerStats)
+	rsp.mutCache.Unlock()
+}
+
+// ComputeRewardsForBlsKey will be called for each BLS key that took part in the consensus (no matter the shard ID)
+// The function will error if something went wrong. It does change the inner state of the called instance.
+func (rsp *rewardsStakingProvider) ComputeRewardsForBlsKey(blsKey []byte) error {
```

Comment:
this interface does not compute rewards and it does not need in the first place to compute. It has to gather data from the staking/auction smart contract. it needs to create 2 maps. One if (BLSkey-ValidatorAddress), the other one if (ValidatorAddress-TotalStaked,TotalNodes), and maybe it can return the TotalAccumulatedStake from all these nodes.

Reply:
Will rename. Regarding the second map, we'll see what we are missing and implement accordingly.

Reply:
Renamed the function to `GetStakingDataForBlsKey`